File: backend/app/models/complaint.py
"""Complaint model for CampusHive."""

from app.database import db
from app.models.base import BaseModel, TimestampMixin


class Complaint(BaseModel, TimestampMixin):
    """complaints table."""

    __tablename__ = 'complaints'

    complaint_id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    organization_id = db.Column(
        db.Integer,
        db.ForeignKey('organizations.organization_id'),
        nullable=False,
    )
    student_id = db.Column(
        db.Integer,
        db.ForeignKey('users.user_id'),
        nullable=False,
    )
    supervisor_id = db.Column(
        db.Integer,
        db.ForeignKey('users.user_id'),
    )
    department_id = db.Column(
        db.Integer,
        db.ForeignKey('departments.department_id'),
        nullable=False,
    )
    title = db.Column(db.String(200), nullable=False)
    description = db.Column(db.Text, nullable=False)
    location = db.Column(db.String(255))
    ml_priority = db.Column(db.Enum('LOW', 'MEDIUM', 'HIGH'))
    final_priority = db.Column(db.Enum('LOW', 'MEDIUM', 'HIGH'))
    status = db.Column(
        db.Enum('PENDING', 'ASSIGNED', 'IN_PROGRESS', 'RESOLVED', 'REOPENED', 'ESCALATED', 'CLOSED'),
        server_default=db.text("'PENDING'"),
    )
    support_count = db.Column(db.Integer, server_default=db.text('0'))
    deadline = db.Column(db.DateTime)
    resolved_at = db.Column(db.DateTime)
    student_feedback = db.Column(db.Enum('SATISFIED', 'NOT_SATISFIED'))

    __table_args__ = (
        db.Index('organization_id', 'organization_id'),
        db.Index('student_id', 'student_id'),
        db.Index('supervisor_id', 'supervisor_id'),
        db.Index('department_id', 'department_id'),
    )

    images = db.relationship('ComplaintImage', backref='complaint', lazy=True, cascade='all, delete-orphan')
    supports = db.relationship('ComplaintSupport', backref='complaint', lazy=True, cascade='all, delete-orphan')
    chat_messages = db.relationship('ComplaintChat', backref='complaint', lazy=True, cascade='all, delete-orphan')
    status_history = db.relationship('ComplaintStatusHistory', backref='complaint', lazy=True, cascade='all, delete-orphan')
    student = db.relationship(
        "User",
        foreign_keys=[student_id],
        backref="complaints_created"
    )
    supervisor = db.relationship(
        "User",
        foreign_keys=[supervisor_id],
        backref="assigned_complaints"
    )
    department = db.relationship(
        "Department",
        back_populates="complaints"
    )
    def to_dict(self):
        return {
            "complaint_id": self.complaint_id,
            "organization_id": self.organization_id,
            "student_id": self.student_id,
            "department_id": self.department_id,
            "supervisor_id": self.supervisor_id,
            "title": self.title,
            "description": self.description,
            "location": self.location,
            "ml_priority": self.ml_priority,
            "final_priority": self.final_priority,
            "status": self.status,
            "support_count": self.support_count,
            "deadline": self.deadline.isoformat() if self.deadline else None,
            "resolved_at": self.resolved_at.isoformat() if self.resolved_at else None,
            "student_feedback": self.student_feedback,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "student_name": self.student.full_name if self.student else None,
            "department_name": self.department.department_name if self.department else None,
        }

    def __repr__(self):
        return f'<Complaint {self.complaint_id}>'

File: frontend/campushive_app/lib/models/complaint.dart
class Complaint {
  const Complaint({
    required this.complaintId,
    required this.organizationId,
    required this.studentId,
    required this.departmentId,
    required this.supervisorId,
    required this.title,
    required this.description,
    required this.location,
    required this.mlPriority,
    required this.finalPriority,
    required this.status,
    required this.supportCount,
    required this.deadline,
    required this.resolvedAt,
    required this.studentFeedback,
    required this.createdAt,
    required this.updatedAt,
    required this.studentName,
    required this.departmentName,
    this.imageUrl,
    this.isSupported = false,
  });

  final int complaintId;
  final int organizationId;
  final int studentId;
  final int departmentId;
  final int? supervisorId;
  final String title;
  final String description;
  final String? location;
  final String? imageUrl;
  final String? mlPriority;
  final String? finalPriority;
  final String status;
  final int supportCount;
  final bool isSupported;
  final DateTime? deadline;
  final DateTime? resolvedAt;
  final String? studentFeedback;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? studentName;
  final String? departmentName;

  factory Complaint.fromJson(Map<String, dynamic> json) {
    return Complaint(
      complaintId: _int(json['complaint_id']),
      organizationId: _int(json['organization_id']),
      studentId: _int(json['student_id']),
      departmentId: _int(json['department_id']),
      supervisorId: _nullableInt(json['supervisor_id']),
      title: _string(json['title']),
      description: _string(json['description']),
      location: _nullableString(json['location']),
      imageUrl: _nullableString(json['image_url']),
      mlPriority: _nullableString(json['ml_priority']),
      finalPriority: _nullableString(json['final_priority']),
      status: _string(json['status']),
      supportCount: _int(json['support_count']),
      isSupported: json['is_supported'] == true,
      deadline: _date(json['deadline']),
      resolvedAt: _date(json['resolved_at']),
      studentFeedback: _nullableString(json['student_feedback']),
      createdAt: _date(json['created_at']),
      updatedAt: _date(json['updated_at']),
      studentName: _nullableString(json['student_name']),
      departmentName: _nullableString(json['department_name']),
    );
  }

  static int _int(Object? value) => value is num ? value.toInt() : 0;

  static int? _nullableInt(Object? value) =>
      value is num ? value.toInt() : null;

  static String _string(Object? value) => value?.toString() ?? '';

  static String? _nullableString(Object? value) {
    final text = value?.toString();
    return text == null || text.isEmpty ? null : text;
  }

  static DateTime? _date(Object? value) =>
      value is String ? DateTime.tryParse(value) : null;
}
